projects/core/src/product/services/product-loading.service.ts

```diff
@@ -0,0 +1,202 @@
+      ),
+      map(
+        productState =>
+          !productState.loading && !productState.success && !productState.error
+      ),
+      distinctUntilChanged(),
+      filter(x => !!x)
+    );
+
+    const isLoading$ = this.store.pipe(
+      select(
+        ProductSelectors.getSelectedProductLoadingFactory(productCode, scope)
+      )
+    );
+
+    const productLoadLogic$ = merge(
+      shouldLoad$,
+      ...this.getProductReloadTriggers(productCode, scope)
+    ).pipe(
+      debounceTime(0),
+      withLatestFrom(isLoading$),
+      tap(([, isLoading]) => {
+        if (!isLoading) {
+          this.store.dispatch(
+            new ProductActions.LoadProduct(productCode, scope)
+          );
+        }
+      })
+    );
+
+    const productData$ = this.store.pipe(
+      select(ProductSelectors.getSelectedProductFactory(productCode, scope))
+    );
+
+    return using(() => productLoadLogic$.subscribe(), () => productData$).pipe(
+      shareReplay({ bufferSize: 1, refCount: true })
+    );
+  }
+
+  /**
+   * Returns reload triggers for product per scope
+   *
+   * @param productCode
+   * @param scope
+   */
+  protected getProductReloadTriggers(
+    productCode: string,
+    scope: string
+  ): Observable<boolean>[] {
+    const triggers = [];
+
+    // max age trigger add
+    const maxAge = this.loadingScopes.getMaxAge('product', scope);
+    if (maxAge && isPlatformBrowser(this.platformId)) {
+      const loadSuccess$ = this.actions$.pipe(
+        ofType(ProductActions.LOAD_PRODUCT_SUCCESS),
+        filter(
+          (action: ProductActions.LoadProductSuccess) =>
+            action.payload.code === productCode && action.meta.scope === scope
+        )
+      );
+
+      const loadStart$ = this.actions$.pipe(
+        ofType(ProductActions.LOAD_PRODUCT),
+        filter(
+          (action: ProductActions.LoadProduct) =>
+            action.payload === productCode && action.meta.scope === scope
+        )
+      );
+
+      triggers.push(this.getMaxAgeTrigger(loadStart$, loadSuccess$, maxAge));
+    }
+
+    return triggers;
+  }
+
+  /**
+   * Generic method that returns stream triggering reload by maxAge
+   *
+   * Could be refactored to separate service in future to use in other
+   * max age reload implementations
+   *
+   * @param loadStart$ Stream that emits on load start
+   * @param loadSuccess$ Stream that emits on load success
+   * @param maxAge max age
+   */
+  private getMaxAgeTrigger(
+    loadStart$: Observable<any>,
+    loadSuccess$: Observable<any>,
+    maxAge: number
+  ): Observable<boolean> {
+    let timestamp = 0;
+
+    const timestamp$ = loadSuccess$.pipe(tap(() => (timestamp = Date.now())));
+
+    const shouldReload$: Observable<boolean> = defer(() => {
+      const age = Date.now() - timestamp;
+
+      const timestampRefresh$ = timestamp$.pipe(
+        delay(maxAge),
+        mapTo(true),
+        withdrawOn(loadStart$)
+      );
+
+      if (timestamp === 0 || age > maxAge) {
```

Comment:
1. Is `timestamp === 0` really needed?
2. Can we explain in a comment why we need this `if`?

Reply:
nope, forget to remove